SG14/algorithm_ext.h

```diff
@@ -80,52 +78,80 @@ namespace stdext
 	template<class BidirIt, class UnaryPredicate>
 	BidirIt unstable_remove_if(BidirIt first, BidirIt last, UnaryPredicate p)
 	{
-		while (1) {
-			while ((first != last) && p(*first)) {
+		while (true) {
+			// Find the first instance of "p"...
+			while (true) {
+				if (first == last) {
+					return first;
+				}
+				if (p(*first)) {
+					break;
+				}
 				++first;
 			}
-			if (first == last--) break;
```

Comment:
I should point out that I'm pretty sure this old line invoked undefined behavior on

    std::vector<int> a;
    unstable_remove(a.begin(), a.end(), 0);

because you're not allowed to decrement `last` beyond the beginning of the vector, not even with postfix `--`.